### ISSUE TITLE
unit ID values leak information

### ISSUE THREAD

**Issue by jjs47** [bwapi/bwapi#799]
It appears that unit IDs are issued sequentially for both players. This means that unit IDs of one player convey information about the other player's production. For example, I have verified by experiment that a terran bot can recognize when a zerg opponent is going 4 pool using only the unit IDs of its own SCVs as they are produced.

I think the natural mitigation is to make IDs random instead of sequential.

Issue noted by IIlIlIIllIlllIlI in comments at http://satirist.org/ai/starcraft/blog/archives/645-BWAPI-and-information-leaks.html

**Comment by N00byEdge** [comment]
This has been a problem for a very long time. It has also been accounted for in the BWAPI 5 design, and should not be relevant anymore after the update is finished.

**Comment by heinermann** [comment]
Yep that is one of the issues being addressed in the new design.

I think we can do more about bots knowing unit orders though, seems like it should be easy to restrict that information especially in cases where the target location/unit isn't visible or is too far away to draw any conclusions. If a SCV isn't *currently* repairing a unit or building something then the other bot shouldn't know that's what it's going to do.

**Comment by chriscoxe** [comment]
Maybe it's not just the id that leaks - maybe the the pointer addresses of the objects of the various types (unit/player/bullet/etc) could also be used to figure out the sequence?  E.g. may need to also randomize the location of type info in data structures such as the unit array.  This was mentioned when talking about units going out and coming into vision in a closed (WontFix) issue: https://github.com/bwapi/bwapi/issues/191

It may impact performance if arrays etc need to often be shuffled though.
